man/linogram.1
.\" -*- NROFF -*-
.\" 
.TH "linogram" "1" "" "Kevin Rosenberg" "Engineering"
.SH "NAME"
linogram \- Linogram gribddisplay for CTSIM
.SH "SYNOPSIS"
.B linogram [OPTIONS]
.SH "DESCRIPTION"
\fIlinogram\fP displays the grid used in linogram\-based scanning for 
the CTSim package. 
.SH "OPTIONS"
\fIlinogram\fP accepts the following options
.TP 12
.B \-\-version      Display version
.TP 12
.B \-\-help         Print a terse help message
.SH "AUTHORS"
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.SH "HISTORY"
CTSim was begun in 1983 using MS\-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.SH "SEE ALSO"
.BR ctsim (1)
.BR ctsimtext (1)
.BR if1 (1)
.BR if2 (1)
.BR ifinfo (1)
.BR ifexport (1)
.BR phm2helix (1)
.BR phm2if (1)
.BR phm2pj (1)
.BR pj2if (1)
.BR pjHinterp (1)
.BR pjinfo (1)
.BR pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/linogram.cpp
/*****************************************************************************
** FILE IDENTIFICATION
**
**   Name:          linogram.cpp
**   Purpose:       Display linogram sampling
**   Programmer:    Kevin Rosenberg
**   Date Started:  April 2003
**
**  This is part of the CTSim program
**  Copyright (C) 1983-2009 Kevin Rosenberg
**
**  This program is free software; you can redistribute it and/or modify
**  it under the terms of the GNU General Public License (version 2) as
**  published by the Free Software Foundation.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program; if not, write to the Free Software
**  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

#include "ct.h"

enum { O_XY, O_POLAR_RT, O_VERBOSE, O_HELP, O_VERSION, O_DEBUG };

static struct option my_options[] =
{
  {"xy", 0, 0, O_XY},
  {"polar-rt", 0, 0, O_POLAR_RT},
  {"debug", 0, 0, O_DEBUG},
  {"verbose", 0, 0, O_VERBOSE},
  {"help", 0, 0, O_HELP},
  {"version", 0, 0, O_VERSION},
  {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";


void
linogram_usage (const char *program)
{
  std::cout << "usage: " << fileBasename(program) << " n d [OPTIONS]\n";
  std::cout << "Imagefile information\n";
  std::cout << std::endl;
  std::cout << "     n            Linogram N\n";
  std::cout << "     d            Max detector spacing\n";
  std::cout << "     --xy         Output x,y pairs\n";
  std::cout << "     --polar-rt   Output r,t pairs\n";
  std::cout << "     --debug      Debug mode\n";
  std::cout << "     --verbose    Verbose mode\n";
  std::cout << "     --version    Print version\n";
  std::cout << "     --help       Print this help message\n";
}

int
linogram_main (int argc, char *const argv[])
{
  int opt_polar_rt = 0;
  int opt_xy = 0;
  int opt_verbose = 0;
  int opt_debug = 0;

  while (1)
    {
      int c = getopt_long (argc, argv, "", my_options, NULL);

      if (c == -1)
        break;

      switch (c)
        {
        case O_XY:
          opt_xy = 1;
          break;
        case O_POLAR_RT:
          opt_polar_rt = 1;
          break;
        case O_VERBOSE:
          opt_verbose = 1;
          break;
        case O_DEBUG:
          opt_debug = 0;
          break;
        case O_VERSION:
#ifdef VERSION
          std::cout << "Version " << VERSION << std::endl << g_szIdStr << std::endl;
#else
          std::cout << "Unknown version number\n";
#endif
          return (0);
        case O_HELP:
        case '?':
          linogram_usage(argv[0]);
          return (0);
        default:
          linogram_usage(argv[0]);
          return (1);
        }
    }

  if (optind + 2 != argc) {
    linogram_usage (argv[0]);
    return (1);
  }

  const char* in_n = argv[optind];
  const char* in_d = argv[optind+1];

  int n = atol (in_n);
  double d = atof (in_d);
  int size = 4 * n + 3;
  int max = 2 * n + 1;
  int min = -max;
  double theta_base = PI/4;
  //  theta_base = 0;

  double theta_vec [size];
  for (int i = 0; i < size; i++) {
    int m = i - (2 * n + 1);
    theta_vec[i] = atan (static_cast<double>(2 * m) / size);
  }

  if (opt_xy) {
    int m;
    for (m = 0; m < size; m++) {
      double step = d * cos(theta_vec[m]);
      for (int id = min; id <= max; id++) {
        double r = id * step;
        double x = r * cos(theta_vec[m] + theta_base);
        double y = r * sin(theta_vec[m] + theta_base);
        printf ("%lf,%lf ", x, y);
      }
      printf ("\n");
    }

    for (m = 0; m < size; m++) {
      double step = d * cos(theta_vec[m]);
      for (int id = min; id <= max; id++) {
        double r = id * step;
        double x = r * cos(theta_vec[m] + PI/2 + theta_base);
        double y = r * sin(theta_vec[m] + PI/2 + theta_base);
        printf ("%lf,%lf ", x, y);
      }
      printf ("\n");
    }
  } else {
    int m;
    for (m = 0; m < size; m++) {
      if (! opt_polar_rt)
        printf ("%lf: ", theta_vec[m] + theta_base);
      double step = d * cos(theta_vec[m]);
      for (int id = min; id <= max; id++) {
        if (opt_polar_rt)
          printf ("%lf,", theta_vec[m] + theta_base);
        printf ("%lf ", id * step);
      }
      printf ("\n");
    }

    for (m = 0; m < size; m++) {
      if (! opt_polar_rt)
        printf ("%lf: ", theta_vec[m] + PI/2 + theta_base);
      double step = d * cos(theta_vec[m]);
      for (int id = min; id <= max; id++) {
        if (opt_polar_rt)
          printf ("%lf,", theta_vec[m] + PI/2 + theta_base);
        printf ("%lf ", id * step);
      }
      printf ("\n");
    }
  }

  return (0);
}

#ifndef NO_MAIN
int
main (int argc, char *const argv[])
{
  int retval = 1;

  try {
    retval = linogram_main(argc, argv);
  } catch (exception e) {
          std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
          std::cerr << "Unknown exception\n";
  }

  return (retval);
}
#endif
